core/Pos/Binary/Class/Primitive.hs

```diff
@@ -240,3 +246,25 @@ decodeUnknownCborDataItem :: D.Decoder s ByteString
 decodeUnknownCborDataItem = do
     decodeCborDataItemTag
     D.decodeBytes
+
+-- | Encodes a type `a` , protecting it from tampering/network-transport-alteration by
+-- protecting it with a CRC.
+encodeCrcProtected :: Bi a => a -> E.Encoding
+encodeCrcProtected x = E.encodeListLen 2 <> encodeUnknownCborDataItem body <> encode (crc32 body)
+  where
+    body = serialize' x
+
+-- | Decodes a CBOR blob into a type `a`, checking the serialised CRC corresponds to the computed one.
+decodeCrcProtected :: forall s a. Bi a => D.Decoder s a
+decodeCrcProtected = do
+    enforceSize ("decodeCrcProtected: " <> show (typeOf (Proxy @a))) 2
+    body <- decodeUnknownCborDataItem
+    expectedCrc  <- decode
+    let actualCrc = crc32 body
+    let crcError = format ("decodeCrcProtected, expected CRC " % shown % " was not the computed one, which was " % shown)
```

Comment:
There is `formatToString` in `formatting`.

Reply:
Wow, there is a combinator for everything in `formatting`, it seems! 😉